pkg/proxy/dns.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/completion"
+	"github.com/cilium/cilium/pkg/fqdn/dnsproxy"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/proxy/logger"
+	"github.com/cilium/cilium/pkg/revert"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// DNSProxyPort is the user-configured global, shared, DNS listen port used
+	// by the DNS Proxy. Both UDP and TCP are handled on the same port. When it
+	// is 0 a random port will be assigned, and can be obtained from
+	// DefaultDNSProxy below.
+	DNSProxyPort int
+
+	// DefaultDNSProxy is the global, shared, DNS Proxy singleton.
+	DefaultDNSProxy *dnsproxy.DNSProxy
+)
+
+// dnsRedirect implements the Redirect interface for an l7 proxy
+type dnsRedirect struct {
+	redirect             *Redirect
+	endpointInfoRegistry logger.EndpointInfoRegistry
+	conf                 dnsConfiguration
+	DNSProxyPort         uint16
+	currentRules         policy.L7DataMap
+}
+
+type dnsConfiguration struct {
+}
+
+// UpdateRules replaces old l7 rules of a redirect with new ones.
+func (dr *dnsRedirect) UpdateRules(wg *completion.WaitGroup) error {
+	for _, rule := range dr.currentRules {
+		for _, dnsRule := range rule.DNS {
+			DefaultDNSProxy.RemoveAllowed(dnsRule.MatchName, fmt.Sprintf("%d", dr.redirect.endpointID))
```

Comment:
This makes the update non-atomic? I.e., it is possible for the dns proxy to see a set of rules that is neither the old or the new set, but some variation in between?

If so, maybe instead prepare a new set of rules and make the new set of rules visible to the proxy via an atomic pointer swap like `setPolicyMap()` in `proxylib/proxylib/instance.go`?